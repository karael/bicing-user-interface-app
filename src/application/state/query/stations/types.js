import { defineAction } from 'redux-define';

import { START, PENDING, CANCELLED, SUCCESS, FAILURE } from 'application/state/stateConstants';

const domain = defineAction('stations');

export const FETCH_LIST = defineAction(
    'FETCH_LIST',
    [START, PENDING, CANCELLED, SUCCESS, FAILURE],
    domain
);
