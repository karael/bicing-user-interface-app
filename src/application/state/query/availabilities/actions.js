import { FETCH } from 'application/state/query/availabilities/types';

export const fetchStart = (
  itineraryStep,
  itineraryAt,
  periodStartAt,
  periodEndAt,
  interval,
  ...stationIds
) => ({
  error: false,
  meta: { isFetching: true, itineraryStep, itineraryAt },
  payload: {
    periodStartAt, periodEndAt, interval, stationIds,
  },
  type: FETCH.START,
});
export const fetchPending = itineraryStep => ({
  error: false,
  meta: { isFetching: true, itineraryStep },
  type: FETCH.PENDING,
});
export const fetchSuccess = (itineraryStep, data) => ({
  error: false,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = (itineraryStep, error) => ({
  error: true,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.FAILURE,
  payload: error,
});
