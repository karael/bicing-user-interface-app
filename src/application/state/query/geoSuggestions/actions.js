import { FETCH } from 'application/state/query/geoSuggestions/types';

export const fetchStart = () => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = geoSuggestions => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: geoSuggestions,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
