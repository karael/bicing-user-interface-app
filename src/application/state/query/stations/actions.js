import { FETCH } from 'application/state/query/stations/types';

export const fetchStart = (latitude, longitude, limit) => ({
  error: false,
  meta: { isFetching: true },
  payload: { latitude, longitude, limit },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = data => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
