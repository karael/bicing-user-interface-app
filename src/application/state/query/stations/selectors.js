const selector = state => state.query.stations;

export const data = state => selector(state).data;
export const error = state => selector(state).error;
export const isFetching = state => selector(state).isFetching;
export const byGeoLocationFilter = state => selector(state).byGeoLocationFilter;
export const stations = (state) => {
  const stationsData = data(state);

  return stationsData && stationsData['hydra:member'] ? stationsData['hydra:member'] : null;
};
