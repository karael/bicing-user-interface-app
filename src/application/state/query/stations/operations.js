import { call, put, takeLatest } from 'redux-saga/effects';

import { fetchListSuccess } from './actions';
import * as Types from './types';
import HttpStationQuery from './../../../../infrastructure/bicingApi/HttpStationQuery';

export function* list(action) {
  try {
    const response = yield call(HttpStationQuery.findAll);
    yield put(fetchListSuccess(response.data));

    // return response.data;
  } catch (e) {
    console.log('LALAL ERROR operations.js', e);
  }
}

export default function* operation() {
  yield takeLatest(Types.FETCH_LIST.START, list);
}
