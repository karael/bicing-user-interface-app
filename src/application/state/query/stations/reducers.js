import { FETCH } from 'application/state/query/stations/types';
import produce from 'immer';
import { createReducer } from 'reduxsauce';

export const INITIAL_STATE = {
  data: undefined,
  error: false,
  isFetching: false,
  latitude: undefined,
  longitude: undefined,
  limit: undefined,
};

export const fetchStart = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.isFetching = action.meta.isFetching;
  draft.latitude = action.payload.latitude;
  draft.longitude = action.payload.longitude;
  draft.limit = action.payload.limit;
});

export const fetchPending = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.isFetching = action.meta.isFetching;
});

export const fetchSuccess = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = action.payload;
  draft.isFetching = action.meta.isFetching;
});

export const fetchFailure = (state = INITIAL_STATE, action) => produce(state, (draft) => {
  draft.data = action.payload;
  draft.error = action.error;
  draft.isFetching = action.meta.isFetching;
});

const HANDLERS = {
  [FETCH.START]: fetchStart,
  [FETCH.PENDING]: fetchPending,
  [FETCH.SUCCESS]: fetchSuccess,
  [FETCH.FAILURE]: fetchFailure,
};

export default createReducer(INITIAL_STATE, HANDLERS);
