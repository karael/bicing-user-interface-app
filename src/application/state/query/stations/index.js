import reducer from './reducers';
import operation from './operations';
import * as actions from './actions';
import * as Types from './types';
import * as selectors from './selectors';

export {
    actions,
    Types,
    operation,
    selectors
};

export default reducer;
