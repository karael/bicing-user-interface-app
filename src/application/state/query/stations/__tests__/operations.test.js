import * as actions from 'application/state/query/stations/actions';
import operation, { fetch } from 'application/state/query/stations/operations';
import { FETCH } from 'application/state/query/stations/types';
import HttpStationQuery from 'infrastructure/bicingApi/HttpStationQuery';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';

describe('application/state/query/stations/operations', () => {
  test('should wait for a fetch start event to fetch with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(FETCH.START, fetch);
  });

  test('should dispatch a fetchListPending action with fetch() generator', () => {
    testSaga(fetch)
      .next()
      .put(actions.fetchPending());
  });

  test('should handle error when byGeoLocationFilter creation is not validated', () => {
    const action = {
      error: false,
      meta: { isFetching: true },
      payload: { latitude: 'bad_value', longitude: 'bad_value', limit: 'bad_value' },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });

  test('should list expected stations with fetch() generator', () => {
    const fakeStations = [
      {
        '@id': '/api/stations/f7fa1d7b-4a7b-410d-bae0-549d862a2523',
        '@type': 'stationView',
        id: 'f7fa1d7b-4a7b-410d-bae0-549d862a2523',
        name: '233 - C/NOU DE LA RAMBLA, 164',
        type: 'BIKE',
        address: 'Nou de la Rambla',
        addressNumber: '164',
        zipCode: '08004',
        latitude: 41.371965,
        longitude: 2.166871,
      },
    ];

    const action = {
      error: false,
      meta: { isFetching: true },
      payload: { latitude: 41.3244, longitude: 2.345, limit: 5000 },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationQuery.findAll), fakeStations],
      ])
      .put(actions.fetchSuccess(fakeStations))
      .run();
  });

  test('should handle error when api call response does not contains expected schema type', () => {
    const fakeStationsWithMissingRequiredProperties = [{ name: '233 - C/NOU DE LA RAMBLA, 164' }];

    const action = {
      error: false,
      meta: { isFetching: true },
      payload: { latitude: 41.3244, longitude: 2.345, limit: 5000 },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationQuery.findAll), fakeStationsWithMissingRequiredProperties],
      ])
      .run()
      .then((result) => {
        expect(result.toJSON()).toMatchSnapshot();
      });
  });

  test('should handle error when api call failed in fetch() generator', () => {
    const error = new Error('error_api_call');
    const action = {
      error: false,
      meta: { isFetching: true },
      payload: { latitude: 41.3244, longitude: 2.345, limit: 5000 },
      type: FETCH.START,
    };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationQuery.findAll), throwError(error)],
      ])
      .put(actions.fetchFailure(error))
      .run();
  });
});
