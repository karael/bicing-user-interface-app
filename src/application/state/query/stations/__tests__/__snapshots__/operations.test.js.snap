// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/stations/operations should handle error when api call response does not contains expected schema type 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "latitude": 41.3244,
            "limit": 5000,
            "longitude": 2.345,
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/find",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stations/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Array [
              Object {
                "name": "233 - C/NOU DE LA RAMBLA, 164",
              },
            ],
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "id",
                  "label": "id",
                },
                "message": "\\"id\\" is required",
                "path": Array [
                  0,
                  "id",
                ],
                "type": "any.required",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/stations/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`application/state/query/stations/operations should handle error when byGeoLocationFilter creation is not validated 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/stations/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Object {
              "latitude": "bad_value",
              "limit": "bad_value",
              "longitude": "bad_value",
            },
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "latitude",
                  "label": "latitude",
                  "value": "bad_value",
                },
                "message": "\\"latitude\\" must be a number",
                "path": Array [
                  "latitude",
                ],
                "type": "number.base",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/stations/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
